architecture/core.rst

```diff
@@ -390,6 +390,183 @@ normal world supports scheduling of processes. Even on UP systems, supporting
 several trusted threads in optee_os helps normal world scheduler to be
 efficient.
 
+Core handlers for native interrupts
+===================================
+
+OP-TEE core offers means to implement native interrupt
+(see :ref:`native_foreign_irqs`) controller drivers and means for device
+drivers to configure and register handler functions for these interrupts.
+Interrupt handlers can be nested for when an interrupt controller
+exposes interrupts which signaling is multiplexed on an interrupt
+controlled by a parent interrupt controller.
+
+Interrupt controllers are represented by an instance of ``struct itr_chip``.
+An interrupt controller exposes a given number of interrupts, each identified
+by an index from 0 to N-1 where N is the total number of interrupts exposed
+by that controller. In the literature, an interrupt index identifier
+is called interrupt number.
+
+The root interrupt controller currently implemented in OP-TEE core is a
+GIC instance for the Arm architecture.
+
+**Interrupt management API functions**
+
+Interrupt management resources are declared in header file
+kernel/interrupt.h_. Main API function are:
+
+    - ``interrupt_enable()`` and ``interrupt_disable()`` to respectively
+      enable and disable and interrupt.
+
+    - ``interrupt_mask()`` and ``interrupt_unmask()`` to respectively
+      mask/unmask and interrupt.
+
+    - ``interrupt_configure()`` to configure an interrupt detection mode
+      and the interrupt priority.
+
+    - ``interrupt_add_handler()`` and ``interrupt_remove_handler()`` to
+      respectively register and unregister an interrupt handler function
+      for a given interrupt.
+
+When configuration switch ``CFG_DRIVERS_INTERRUPT`` is enabled, core
+offers a means to use Device Tree to describe interrupts topologyin the
+platform. When used, the platform must provide a secure devicetree blob
+(DTB) which platform and device drivers can use to getreferences to
+interrupt controllers and relative interrupt numbers theyconsume.
+Main API functions are:
+
+    - ``dt_register_interrupt_provider()`` to register an interrupt
+      controller related to a given node in the Device Tree. This function
```

Comment:
Shouldn't the function be called `dt_register_interrupt_controller()` then?

Reply:
I understand your point of view. However all functions used the register a provider of device driver reference in _dt_drvier.c_ (and friends) resources are labelled `dt_register_xxx_provider()` so I would prefer to keep that `_provider` suffix. Here I could propose `dt_register_interrupt_controller_provider()` but it's a bit long, almost as long  as existing `dt_driver_device_from_node_idx_prop_phandle()` which I feel pretty bad about :(

What about `dt_register_itr_chip_provider()`?

Reply:
Ah, consistency again ;) it's fine to keep `dt_register_interrupt_provider()` then.